Documentation/proposal-runconfig_inside_osv.md

```diff
@@ -1,55 +1,286 @@
+* set run command to the existing OSv image
+
+Other infrastructures (OpenStack, Amazon) are unable to perform first and second bullet, but can do
+the third one.
+
+### Executive summary
+This proposal is about what actions will Capstan perform on COMPOSE time based on user-friendly
+input file that is provided in `$CWD/meta/run.yaml`. The format of this file was decided [here](https://github.com/mikelangelo-project/capstan/issues/3#issuecomment-245210949) and is not to be changed much with this proposal
+(values starting with $ will be parsed as placeholder name, that's the only difference).
+
+The idea is that Capstan would prepare everything in COMPOSE time so that the first and second bullet
+(see Problem description section) would not be needed at the RUN time. This way OpenStack would be
+able to make use of named runtimes by not even having Capstan installed.
 
 I propose the following solution:
 
+### User aspect
+OSv packaging with runtime support ends up with four user roles:
+
+##### Application user
+Downloads a pre-composed image and RUN it without modifications. She uses EXTERNAL packages and
+EXTERNAL application. She does not perform COMPOSE step. She must be provided with this information:
+ * what configuration names are available
+ * what parameters are needed for each configuration name
+
+Example: Hdfs. User downloads it and runs it with customized parameters.
+
+##### Runtime user
+Downloads pre-prepared Capstan packages. She uses EXTERNAL packages and INTERNAL application. She
+needs to provide the `meta/run.yaml` file. There she needs to pick one of the supported runtimes that suites her best. Then she needs to either provide runtime-specific values either define placeholder
+names for them (explained in following sections). She creates arbitrary number of configurations and
+names them.
+
+To successfully perform the COMPOSE step, she must be provided with this information:
+ * what runtimes are available
+ * what runtime-specific attributes are supported
+
+Once she successfully composes the image, she takes the role of Application user.
+
+Example: NodeJS. User uses 'node' runtime (which downloads pre-prepared node package), but provides
+her own NodeJS JavaScript application code.
+
+##### Runtime developer
+Extends Capstan tool by implementing runtime.Runtime interface (Golang programmer). She must be
+provided with this information:
+ * what packages are available
+ * how to use package
+
+It must be as easy as possible to add new runtime since we may expect PRs for new runtimes!
+
+QUICK THOUGHT: Is it really neccessary that runtime is hard-coded in Capstan? Would it be better to
+support runtimes written in yaml and included in package itself? Not sure this is possible, though.
+
+##### Package developer
+Statically compiles new languages from source into shared objects and bundles them into Capstan
+package.
+
+NOTE: There is a mess with the hierarchy of packages. Would it make sense to introduce "package type"?
```

Comment:
Hmmm, I don't think that's necessary right now. Imagine the user would simply compose the app with Node.JS (runtime) and OSv's HTTP server. Users of this VM image would boot an instance and then use HTTP REST API to upload their javascript sources. After the upload, they would just run the node.